src/Raven.Server/Config/JsonConfigFileModifier.cs

```diff
@@ -0,0 +1,225 @@
+        {
+            throw new UnsuccessfulFileAccessException(e, path, fileAccess);
+        }
+    }
+
+    private async Task PersistConfiguration(BlittableJsonReaderObject json)
+    {
+        using var tempFile = new TempFile(_path);
+
+        await using (var file = OpenFile(tempFile.Path, FileMode.Create, FileAccess.ReadWrite))
+        await using (var streamWriter = new StreamWriter(file))
+        await using (var writer = new JsonTextWriter(streamWriter))
+        await using (var reader = new BlittableJsonReader())
+        {
+            writer.Formatting = Formatting.Indented;
+            reader.Initialize(json);
+
+            await writer.WriteTokenAsync(reader);
+                
+            await writer.FlushAsync();
+            await streamWriter.FlushAsync();
+            file.Flush(true);
+        }
+
+        Validate(tempFile.Path);
+        SwitchTempWithOriginalAndCreateBackup(tempFile.Path);
+    }
+
+    private void SwitchTempWithOriginalAndCreateBackup(string tempPath)
+    {
+        try
+        {
+            if (File.Exists(_path))
+            {
+                File.Replace(tempPath, _path, _path + ".bak");
+            }
+            else
+            {
+                File.Move(tempPath, _path);
+            }
+            
+            if (PlatformDetails.RunningOnPosix)
+                Syscall.FsyncDirectoryFor(_path);
+        }
+        catch (Exception e) when (e is UnauthorizedAccessException or SecurityException)
+        {
+            throw new UnsuccessfulFileAccessException(e, _path, FileAccess.Write);
+        }
+    }
+
+    private class TempFile : IDisposable
+    {
+        public TempFile(string path)
+        {
+            Path = path + ".tmp";
+        }
+
+        public string Path { get; }
+
+        public void Dispose()
+        {
+            try
+            {
+                File.Delete(Path);
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+    }
+
+    public void Dispose()
+    {
+        _json?.Dispose();
+    }
+}
+
+public class SettingsJsonModifier : JsonConfigFileModifier
+{
+    private SettingsJsonModifier(JsonOperationContext context, string path, bool overwriteWholeFile = false) 
+        : base(context, path, overwriteWholeFile)
+    {
+    }
+
+    public new static SettingsJsonModifier Create(JsonOperationContext context, string path, bool reset = false)
+    {
+        var obj = new SettingsJsonModifier(context, path, reset);
+        obj.Initial();
+        return obj;
+    }
+
+    protected override void Validate(string path)
+    {
+        RavenConfiguration.CreateForTesting("for-validattion", ResourceType.Server, path).Initialize();
+    }
+
+    public void SetOrRemoveIfDefault<T, T1>(T1 value, Expression<Func<RavenConfiguration, T>> getKey)
+    {
+        if (CheckIfDefault(value, RavenConfiguration.GetDefaultValue(getKey)))
```

Comment:
I do not like this, because if someone uses default in the file, you will remove them?

Reply:
done